### PR TITLE
introduce fault items

### PR DESCRIPTION
**Pull request by ryanfkeepers**
Adds the item struct to the fault package for tracking serializable and dedupliatable error sources.

---

#### Does this PR need a docs update or release note?

- [x] :no_entry: No

#### Type of change

- [ ] :sunflower: Feature

#### Issue(s)

* #2708

#### Test Plan

- [x] :zap: Unit test

### CODE REVIEW

File: src/pkg/fault/item.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
+package fault
+
+type itemKind string
```
Comment (dadams39) [inline src/pkg/fault/item.go:3]: I understand that we already have `itemType`. However, it does seem to work in this connotation. There are other things that may be more useful:
- class
- division
- group

File: src/pkg/fault/item.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+const (
+	ItemKindFile          itemKind = "file"
+	ItemKindContainer     itemKind = "container"
+	ItemKindResourceOwner itemKind = "resource_owner"
```
Comment (dadams39) [inline src/pkg/fault/item.go:8]: It's unclear why `Kind` is required within the constant name. 
- `ItemFile`
- `ItemContainer`

flows better or `GroupFile`, `CatFile`

File: src/pkg/fault/item.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+)
+
+// in pkg fault.  eg: fault.Item{}
```
Comment (dadams39) [inline src/pkg/fault/item.go:11]: more comments needed for the struct

File: src/pkg/fault/item.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+type Item struct {
+	// deduplication identifier; the ID of the item under scrutiny.
```
Comment (dadams39) [inline src/pkg/fault/item.go:13]: `ID of the observed item`

File: src/pkg/fault/item.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	ID string `json:"id"`
+
+	// a human-readable reference: file/container name, email, etc
+	Name string `json:"name"`
+
+	// the name and id of the container inside which this item is
+	// stored.  May be empty if the item is not stored within a
+	// container (ex: resourceOwner)
```
Comment (dadams39) [inline src/pkg/fault/item.go:21]: This is comment is unclear.

File: src/pkg/fault/item.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	ContainerName string `json:"containerName"`
+	ContainerID   string `json:"containerID"`
+
+	// tracks the type of item represented by this entry.
+	Kind itemKind `json:"kind"`
```
Comment (dadams39) [inline src/pkg/fault/item.go:26]: Classification?

File: src/pkg/fault/item.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+	// Error() of the causal error, or a sentinel if this is the
+	// source of the error.  In case of ID collisions, the first
+	// item takes priority.
+	Cause string `json:"cause"`
```
Comment (dadams39) [inline src/pkg/fault/item.go:31]: Love this

File: src/pkg/fault/item.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	if len(i.Kind) == 0 {
+		return "processing item of unknown kind"
+	}
+
+	return string("processing " + i.Kind)
+}
+
+// ContainerErr produces a Container-kind Item.
```
Comment (dadams39) [inline src/pkg/fault/item.go:47]: produces or returns?

File: src/pkg/fault/item.go (hunk 2 above)
Comment (ryanfkeepers) [reply src/pkg/fault/item.go:8]: Just a reference to its underlying type, and a way to namespace it from other similarly named consts.

File: src/pkg/fault/item.go (hunk 1 above)
Comment (ryanfkeepers) [reply src/pkg/fault/item.go:3]: * `class` too overloaded a term in software
* `division` makes me think that this is a divider, which it is not.
* `group` could work, but imo it reads identically to `kind`.

File: src/pkg/fault/item.go (hunk 8 above)
Comment (ryanfkeepers) [reply src/pkg/fault/item.go:47]: constructs 😁

File: src/pkg/fault/item.go (hunk 6 above)
Comment (ryanfkeepers) [reply src/pkg/fault/item.go]: swapped `kind` for `type`.

**Review by dadams39 — APPROVED** [review]

**Pull request merged**